langchain4j-core/src/main/java/dev/langchain4j/rag/content/retriever/WebSearchContentRetriever.java

```diff
@@ -0,0 +1,114 @@
+import dev.langchain4j.store.embedding.inmemory.InMemoryEmbeddingStore;
+import dev.langchain4j.web.search.WebSearchResults;
+import dev.langchain4j.model.embedding.EmbeddingModel;
+import dev.langchain4j.rag.content.Content;
+import dev.langchain4j.rag.query.Query;
+import dev.langchain4j.store.embedding.EmbeddingMatch;
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import dev.langchain4j.web.search.WebSearchEngine;
+import lombok.Builder;
+
+import java.util.List;
+
+import static dev.langchain4j.internal.Utils.getOrDefault;
+import static dev.langchain4j.internal.ValidationUtils.*;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * A {@link ContentRetriever} that retrieves relevant {@link Content} from the web using a {@link WebSearchEngine}.
+ * <br>
+ * By default, this retriever returns complete web pages: one {@link Content} for each web page that a {@link WebSearchEngine} has returned for a given {@link Query}. For some use cases, this might be suboptimal, as complete web pages can contain too much irrelevant content. For example, for the query "Who painted the Mona Lisa?", Google returns a Wikipedia page that contains everything about the Mona Lisa. For such use cases, this retriever can be configured with a {@link DocumentSplitter} and an {@link EmbeddingModel}, which will be used to split web pages into smaller segments and find the most relevant ones, returning only those.
+ * <br>
+ * <br>
+ * Configurable parameters (optional):
+ * <ul>
+ *     <li>{@link DocumentSplitter} - To split the web pages into smaller text segments.</li>
+ *     <li>{@link EmbeddingModel} - To embed segments and find the most
+ *     relevant ones to the given {@link Query}</li>
+ *     <li>{@link #maxTextSegments} - The maximum number of text segments to return.
+ *     <br>
+ *     Default value is 3.
+ *     </li>
+ * </ul>
+ */
+public class WebSearchContentRetriever implements ContentRetriever{
+
+    private static final int DEFAULT_MAX_TEXT_SEGMENTS = 3;
+    
+    private final WebSearchEngine webSearchEngine;
+    private final DocumentSplitter documentSplitter;
+    private final EmbeddingModel embeddingModel;
+    private final Integer maxTextSegments;
+
+
+    /**
+     * Constructs a new WebSearchContentRetriever with the specified web search engine.
+     * @param webSearchEngine The web search engine to use for retrieving web search results.
+     */
+    public WebSearchContentRetriever(WebSearchEngine webSearchEngine) {
+        this(webSearchEngine, null);
+    }
+
+    /**
+     * Constructs a new WebSearchContentRetriever with the specified web search engine and document splitter.
+     * @param webSearchEngine The web search engine to use for retrieving web search results.
+     * @param documentSplitter The document splitter to use for splitting web search results into smaller text segments.
+     */
+    public WebSearchContentRetriever(WebSearchEngine webSearchEngine, DocumentSplitter documentSplitter) {
```

Comment:
On a third thought 😅 
I wrongly assumed that web search results will contain the whole text of the pages, but it seems to be not the case, right?
They will contain only small snippets? In such a case I think we should make it possible to (optionally, if user configures) fetch/scrap? contents of web pages, ideally in parallel from all web pages/results, something similar to what we have in `DefaultRetrievalAugmentor` with `Executor`. In such case splitter/embedding model make sense, otherwise - no.
If this is too big of a change for this PR and you want to make it as a separate PR or not at all, then I propose to remove splitting/embedding logic from this class for the moment as this does not make any sense to split/embed small snippets.
Sorry about that!

Reply:
Yes it contain only snippets. The traditional web searchers (google, bing, yahoo, etc) already have the intelligence to return the most relevant snippet of the page associated with the user query.
However, they have the problem of being very short and giving the impression that information is missing. I guess that was detected by Tavily AI, which complements a traditional web search with AI and scraped the whole page and rewrites more information in the `content` field.

Following you, we can do the same by using the URI and fetch/scrap the whole page, (perhaps using the `UrlDocumentLoader` or any URL document, suggest me one) and adding the splitter/embedding and the minScore to build a content enriched. I'd use `content` field to do that.

For now, this first version will support only the snippet and in another PR I will implement the content. that's make sense for you?

Reply:
Yeah, totally! Let's roll this first version with just snippets and second version with optional scrapping, splitting, retrieving most relevant parts.

I am not sure it will be 100% reliable to use only `UrlDocumentLoader` as many web sites have scrapping blocks, so we will probably need to find a better solution or just ignore pages which failed to be loaded with `UrlDocumentLoader`...